android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/MainActivity.java

```diff
@@ -83,28 +80,91 @@ public boolean onOptionsItemSelected(MenuItem item) {
     return super.onOptionsItemSelected(item);
   }
 
+  /**
+   * Ensures Bluetooth is available on the beacon and it is enabled. If not,
+   * displays a dialog requesting user permission to enable Bluetooth.
+   */
+  private void ensureBluetoothIsEnabled(BluetoothAdapter bluetoothAdapter) {
+    checkingPermissions = true;
+    if (!bluetoothAdapter.isEnabled()) {
+      Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
+      startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
+      return;
+    }
+    ensureLocationPermissionIsEnabled();
+  }
+
+  @Override
+  protected void onActivityResult (int requestCode, int resultCode, Intent data) {
+    Log.d(TAG, "onActivityResult");
+    if (requestCode == REQUEST_ENABLE_BT && resultCode == -1) {
+      ensureLocationPermissionIsEnabled();
+      return;
+    }
+    Toast.makeText(this, getString(R.string.bt_on), Toast.LENGTH_LONG).show();
+    finish();
+  }
+
   @Override
   protected void onResume() {
     super.onResume();
-    Log.d(TAG, "resumed MainActivity");
-    BluetoothManager btManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
-    BluetoothAdapter btAdapter = btManager != null ? btManager.getAdapter() : null;
-    if (btAdapter == null) {
-      Toast.makeText(getApplicationContext(),
-              R.string.error_bluetooth_support, Toast.LENGTH_LONG).show();
-      finish();
-      return;
+    if(!checkingPermissions){
+      Log.d(TAG, "resumed MainActivity");
+      BluetoothManager btManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
+      BluetoothAdapter btAdapter = btManager != null ? btManager.getAdapter() : null;
+      if (btAdapter == null) {
+        Toast.makeText(getApplicationContext(),
+            R.string.error_bluetooth_support, Toast.LENGTH_LONG).show();
+        finish();
+        return;
+      }
+      if (checkIfUserHasOptedIn()) {
+        ensureBluetoothIsEnabled(btAdapter);
+      } else {
+        // Show the oob activity
+        Intent intent = new Intent(this, OobActivity.class);
+        startActivity(intent);
+      }
     }
-    if (checkIfUserHasOptedIn()) {
-      ensureBluetoothIsEnabled(btAdapter);
-      showNearbyBeaconsFragment();
-      Intent intent = new Intent(this, ScreenListenerService.class);
-      startService(intent);
-    } else {
-      // Show the oob activity
-      Intent intent = new Intent(this, OobActivity.class);
-      startActivity(intent);
+    
+  }
+
+  private void finishLoad() {
+    showNearbyBeaconsFragment();
+    Intent intent = new Intent(this, ScreenListenerService.class);
+    startService(intent);
+  }
+
+  private void ensureLocationPermissionIsEnabled() {
+    if (Build.VERSION.SDK_INT >=23 && ContextCompat.checkSelfPermission(getApplicationContext(),android.Manifest.
```

Comment:
There should be a space between a binary comparison operator and each of its operands:
  Build.VERSION.SDK_INT >= 23
Function parameters listed on the same line should have a space after the comma:
  getApplicationContext(), android.Manifest.
Also, try to insert line breaks at the highest syntactic level possible.  In this case that means breaking after "getApplicationContext()," so that android.Manifest.permission.ACCESS_COARSE_LOCATION can appear on a single line.